## Benchmarking RNA-seq mapping (real data) 

Contains the scripts used for analysing and comparing RNA-seq alignments from *STAR* and *vg mpmap*.

Each script contains hard-coded filenames, however, many of these names are identical to the output names used in the preceding steps in the benchmarking pipeline. Note that the number of threads is set to either 4 or 8 in some of the bash scripts. 

Follow the below steps to replicate the analyses presented in the paper. 

### 1. Calculate general mapping statistics

Scripts (`illumina` folder):

1. `map_stats.sh`: Calculates general mapping statistics stratified by mapping quality. Run script for each method, reference and read set combination.

Command line arguments:

* The prefix for the alignment file (BAM format) is given as a command line argument.

Input files:

* *empty.bed*: Empty BED file (script needs a BED file as input).

Methods:

* [vgrna_scripts](https://github.com/jonassibbesen/hprc-rnaseq-analyses-scripts/tree/main/evaluation/vgrna_scripts)

### 2. Calculate Iso-Seq exon coverage statistics 

Scripts (`iso_seq` folder):

1. `filter.sh`: Merge, sort, rename contigs and filter Iso-Seq alignments. 
2. `exons.sh`: Create exon regions defined by the filtered Iso-Seq alignments. 
3. `coverage.sh`: Calculate exon coverages from the filtered Iso-Seq alignments. 

Input files:

* *ENCFF247TLH.bam*, *ENCFF431IOE.bam*, *ENCFF520MMC.bam*, and *ENCFF626GWM.bam*: Iso-Seq alignment files (4 replicates). See [data.txt](https://github.com/jonassibbesen/hprc-rnaseq-analyses-scripts/blob/main/evaluation/real/mapping/iso_seq/ENCSR706ANY/data.txt) file.

Methods:

* [samtools](https://github.com/samtools/samtools) (v1.15)
* [bedtools](https://github.com/arq5x/bedtools2) (v2.30.0)
* [vgrna_scripts](https://github.com/jonassibbesen/hprc-rnaseq-analyses-scripts/tree/main/evaluation/vgrna_scripts)

### 3. Calculate mapping coverage statistics

Scripts (`illumina` folder):

1. `cov_eval.sh`: Sort alignments and calculate exon coverages from the alignments. Run script for each method, reference and read set combination.

Command line arguments:

* The prefix for the alignment file (BAM format) is given as a command line argument.

Input files:

* *ENCSR706ANY_primary_mapq30.bed*: Exon regions defined by the Iso-Seq alignments generated by the `exons.sh` script in the previous step.

Methods:

* [samtools](https://github.com/samtools/samtools) (v1.15)
* [vgrna_scripts](https://github.com/jonassibbesen/hprc-rnaseq-analyses-scripts/tree/main/evaluation/vgrna_scripts)

### 4. Plot mapping benchmark

Scripts used to plot the results presented in the paper:

* `plot_map_rate_hprc_r1.R`: Creates mapping rate plots.
* `plot_cov_cor_hprc_r1.R`: Creates coverage correlation plots.

Please note that the R scripts assume that the data is structured in a specific set of folders. They might therefore require a bit of reworking before they can be used in other environments. 
